### PR TITLE
snap-packaging for gitea

### PR DESCRIPTION
**Pull request by tbraeutigam**
As per @kwill 's request in  https://github.com/tbraeutigam/gogs-snap/issues/1 and #2539, I modified my gogs-snap to fit with gitea.

**Changes to Gitea Code**:
 * None for actual code
 * `.gitignore`:
   Adding a couple of exceptions for snapcraft files that might come up during compilation.

**New Files**:
 * `snap/snapcraft.yaml`
   The snap package instructions
 * snap/helpers/simple_launcher.sh
   A simplistic wrapper for gitea, to fit with the snap environment
 * snap/helpers/directorySetup.sh
   Sets up the directories/copied data from r/o `$SNAP` to `$SNAP_DATA` and `$SNAP_COMMON`
 * snap/helpers/snapApp.ini
   A few default settings (really 99% of changes are to adjust the paths properly for the Snap environment).

**Rationale**:
A snap version allows for quick testing and/or easy local hosting on your own machine/own network.
The snap does not restrict you from using any of the database options. It defaults to Sqlite3 on the setup page as that is what can be immediately used without further issues from within the snap.

**Tested Scenarios**:
 * Ran through a couple of coming use cases (create repo, issues, clone/pull/push, PR, etc)
 * Builds fine from current master (Tested on Ubuntu 17.04 and 16.04 LTS)

If you do accept the pull-request it would be great if you could set up gitea to be automatically build via https://build.snapcraft.io -- it will allow it to be built on any commit to the master branch (or any other branch you specify).

### CODE REVIEW

**Comment by sapk** [comment]
Is it possible to put the snap folder under contrib/ ?

**Comment by tbraeutigam** [comment]
I've never tried (it's certainly not the norm, and used to be that the
contents were in the root). But as it otherwise allows arbitrary source
paths, it should be an option.

It definitively won't work with the automated build service, though.

If having the snap package information in the root is undesired, the more
common option would be to hold them in a separate repository - this would
take away from the automation again, as only commits to the repository
containing the snap info would be monitored.

File: snap/helpers/app.ini
Comment (jonasfranz) [inline snap/helpers/app.ini:2]: Should gitea really run as root? Or do I misinterpret this variable?

File: snap/helpers/directorySetup.sh
Comment (jonasfranz) [inline snap/helpers/directorySetup.sh]: Why is it requried to run Gitea as root? (I am not familiar with snap-things)

File: snap/helpers/app.ini
Comment (tbraeutigam) [reply snap/helpers/app.ini:2]: It's not a misinterpretation.
Snap packages right now all 'run as root' for daemons.

The security part here is defined via apparmor confinement. With the plugs in snapcraft.yaml the root user is only allowed to access the snap directories and network access and port binding.

File: snap/helpers/directorySetup.sh
Comment (tbraeutigam) [reply snap/helpers/directorySetup.sh]: There could be a way to run it as a user of started manually. While working on the gigs snap though, that got messy very fast. This I stepped that out to be revisited at a later point.

Comment (tbraeutigam) [reply snap/helpers/directorySetup.sh]: Sorry, typing on mobile:
if started*
s/gigs/gogs/
Thus I backed that*

**Comment by tbraeutigam** [comment]
Hi,
I just pushed a few updates to the PR that streamline the wrapper around gitea and move install-time code to the install-hook of snapd.

This moves the `snap/helpers/directorySetup.sh` to `snap/hooks/install` and removes it from being run on every call of the `gitea` binary.

Hope my prior comments and this commit help clean and clear it up a bit.

File: snap/helpers/app.ini
Comment (jonasfranz) [reply snap/helpers/app.ini:2]: That's all right then.

**Comment by tbraeutigam** [comment]
Is there anything I can do for getting approval?

**Comment by lunny** [comment]
@tbraeutigam since not all of the maintainers are familiar with snap-packaging, I think it needs some time.

**Comment by tbraeutigam** [comment]
@lunny That's perfectly valid. If there is anything that I can help clarify/assist with, I'm more than happy to.

**Comment by lunny** [comment]
truested LGTM @tbraeutigam

File: snap/helpers/app.ini
Comment (sapk) [inline snap/helpers/app.ini]: CERT_FILE and KEY_FILE aren't need if protocol is http. Maybe we could generate some at install and default to https ?

**Review by sapk — APPROVED** [review]
Just a little comment. 
For the rest, I will trust @tbraeutigam

File: snap/helpers/app.ini
Comment (tbraeutigam) [reply snap/helpers/app.ini]: A fair bit of time has gone by since the writing and a lot of new features have come to snapcraft, so I'll re-check this. These entries were added back then because I judged including a full config-writer or something hack-ish would've been more costly than having these two lines in there.

Comment (tbraeutigam) [reply snap/helpers/app.ini]: I've kept them in, and in a way 'made it worse' by moving to the snap hooks for configuration.

Doing that I'm using the defaults from app.ini.sample, overwriting the parts necessary for a snap to function and adding them all to the app.ini as used by gitea.

I personally prefer this, as it's now possible to configure everything via `sudo snap set category-subcategory.option`, but it has some drawbacks at the moment, as I didn't want to touch the config system:

* Cateogories in the style of `category.subcategory` are now `category-subcategory`
* Options containing an underscore have a '99' in its place (`ssh.minimum_key_size` is `ssh-minimum99key99size` for snapctl)
* New defaults do not apply automatically
* Service-restart takes longer (it re-parses the whole app.ini every time)

All options can still be changed manually in the app.ini file and will be added to `snapctl` upon service-restart.

Not sure which way is preferred, but unless we'll touch the config system for this, there's not much that can be done (aside from completely leaving it alone and up to the user to set everything).

I would love to have it fully integrated into the snap config system as an option, but that would require a lot of work:

* Category names are only allowed to be lower case and can only contain dashes
* There is no programmatic interface from what I can see, all has to be done with `exec()` calls.

ini files are regrettably hard to convert to the json-based snapctl config format.
(You can see some of the mess by looking at the `configuration.sh` files `ToSnap` and `ToIni` functions...)

**Comment by bkcsoft** [comment]
Why isn't this a separate repo like `gogs-snap` and `gitea-homebrew` ? 🤔

**Comment by bkcsoft** [comment]
Is there any way to trigger the snapcraft buildservers from Drone? 🤔 

EDIT: Was already answered in the issue https://github.com/go-gitea/gitea/issues/2539#issuecomment-361515394
> To get it onto build.snapcraft.io, you just have to go to the site, create an account and log in, and it guides you on how to allow the site to set a trigger for building.

**Comment by lunny** [comment]
@bkcsoft It seems snap don't support that usage? @tbraeutigam

**Comment by lofidevops** [comment]
@bkcsoft , in response to your query:

> Why isn't this a separate repo like gogs-snap and gitea-homebrew ?

An earlier comment by @tbraeutigam :

> [Putting snap package information in a separate repo] would take away from the automation again, as only commits to the repository containing the snap info would be monitored.

File: snap/helpers/directorySetup.sh
Comment (kovariance) [reply snap/helpers/directorySetup.sh]: Hi @tbraeutigam, sorry for bringing this up 2 years after the PR, but do you recall exactly which issue prevented `gitea web` from running correctly without daemon mode?

**Comment by tbraeutigam** [comment]
Hi @pokemaster7,
I believe the issue was with Daemon-Integration in snapd not dealing well
with user-level systemd or similar...
But I *could* be wrong at this time.

**Pull request merged**
